crates/solidity/outputs/cargo/slang_solidity/src/resolver.rs

```diff
@@ -0,0 +1,85 @@
+use metaslang_bindings::PathResolver;
+
+pub struct SolidityPathResolver;
+
+impl PathResolver for SolidityPathResolver {
+    fn resolve_path(&self, context_path: &str, path_to_resolve: &str) -> Option<String> {
+        if is_relative_path(path_to_resolve) {
+            // Relative import: the actual path will be computed using the
+            // context path (ie. the path of the importing source unit)
+            normalize_path(path_to_resolve, get_parent_path(context_path))
+        } else {
+            // Direct import: this path will be used as-is
+            // TODO: implement the VFS concept and use a resolver callback if
+            // the path is not already present
+            Some(path_to_resolve.to_string())
+        }
+    }
+}
+
+fn is_relative_path(path: &str) -> bool {
+    path.starts_with("./") || path.starts_with("../")
+}
+
+fn get_parent_path(path: &str) -> &str {
+    let sep_index = path.rfind('/').map_or(0usize, |index| index + 1);
+    &path[..sep_index]
+}
+
+fn normalize_path(path: &str, base_path: &str) -> Option<String> {
```

Comment:
instead of doing string operations here, I wonder if we can just use `canonicalize()` and `is_absolute()`?

https://doc.rust-lang.org/stable/std/fs/fn.canonicalize.html
https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.is_absolute

Basically something like:

```rust
assert!(!import.is_absolute());
return current_dir.join(import).canonicalize();
```

Reply:
`canonicalize()` is definitely a no go because it does I/O, and the path have to actually exist in the filesystem.

`is_absolute()` _could_ work I guess, although the semantics are different from what we need. The implementation does resolve only relative imports according to [this definition](https://docs.soliditylang.org/en/v0.8.26/path-resolution.html#relative-imports). Eg. `foo.sol` and `./foo.sol` are both relative paths, but for the purposes of Solidity they would be resolved differently. The former is an absolute path for the Solidity VFS.

Reply:
My understanding is that is only used for our tests, where we are always using relative files, so I/O is fine?

Reply:
Sure, but the files are virtual and defined by the separators in each `input.sol`, eg.

```
// --- path: foo.sol

// --- path: lib/bar.sol
```

So the actual files don't exists in the filesystem (there's only `input.sol`).